File: Cycle.cs
using System;
using System.Collections.Generic;

namespace LoopW;

/// <summary>
/// Describes the action selected for one cycle invocation. The cursor is only
/// committed after the window action succeeds, so a failed placement does not
/// consume a step.
/// </summary>
internal readonly record struct CycleSelection(
    WindowAction RequestedAction,
    WindowAction EffectiveAction,
    int Position,
    int Count,
    bool IsCycle)
{
    public string StatusSuffix => IsCycle ? $"  ·  Cycle {Position + 1}/{Count}" : string.Empty;
}

internal static class WindowCycleService
{
    private sealed record Cursor(WindowAction RequestedAction, int Position);

    private static readonly Dictionary<IntPtr, Cursor> Cursors = new();

    public static bool CanCycle(WindowAction action) => TryGetChain(action, out _);

    public static CycleSelection Select(IntPtr window, WindowAction requestedAction, bool enabled)
    {
        if (!enabled || !TryGetChain(requestedAction, out var chain))
        {
            Cursors.Remove(window);
            return new CycleSelection(requestedAction, requestedAction, 0, 0, false);
        }

        var requestedPosition = PositionOf(chain, requestedAction);
        if (requestedPosition < 0)
        {
            requestedPosition = 0;
        }

        var position = requestedPosition;
        if (Cursors.TryGetValue(window, out var cursor) && cursor.RequestedAction == requestedAction)
        {
            position = (cursor.Position + 1) % chain.Count;
        }

        return new CycleSelection(requestedAction, chain[position], position, chain.Count, true);
    }

    public static void Commit(IntPtr window, CycleSelection selection)
    {
        if (selection.IsCycle)
        {
            Cursors[window] = new Cursor(selection.RequestedAction, selection.Position);
        }
    }

    private static int PositionOf(IReadOnlyList<WindowAction> chain, WindowAction action)
    {
        for (var i = 0; i < chain.Count; i++)
        {
            if (chain[i] == action)
            {
                return i;
            }
        }

        return -1;
    }

    private static bool TryGetChain(WindowAction action, out IReadOnlyList<WindowAction> chain)
    {
        chain = action switch
        {
            WindowAction.LeftHalf or WindowAction.LeftThird or WindowAction.LeftTwoThirds =>
                new[] { WindowAction.LeftHalf, WindowAction.LeftThird, WindowAction.LeftTwoThirds },
            WindowAction.RightHalf or WindowAction.RightThird or WindowAction.RightTwoThirds =>
                new[] { WindowAction.RightHalf, WindowAction.RightThird, WindowAction.RightTwoThirds },
            WindowAction.TopHalf or WindowAction.TopThird or WindowAction.TopTwoThirds =>
                new[] { WindowAction.TopHalf, WindowAction.TopThird, WindowAction.TopTwoThirds },
            WindowAction.BottomHalf or WindowAction.BottomThird or WindowAction.BottomTwoThirds =>
                new[] { WindowAction.BottomHalf, WindowAction.BottomThird, WindowAction.BottomTwoThirds },
            _ => Array.Empty<WindowAction>()
        };

        return chain.Count > 0;
    }
}

File: src/core/cycle.rs
//! Directional cycle chains (half -> third -> two-thirds).
//!
//! Mirrors `WindowCycleService`: repeating the same requested action advances
//! through its chain. The cursor only advances after a successful placement,
//! so callers must invoke [`CycleState::commit`] on success.

use std::collections::HashMap;

use super::actions::WindowAction;

/// One cycle step selection.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CycleSelection {
    pub requested: WindowAction,
    pub effective: WindowAction,
    pub position: usize,
    pub count: usize,
    pub is_cycle: bool,
}

impl CycleSelection {
    pub fn status_suffix(self) -> String {
        if self.is_cycle {
            format!("  ·  Cycle {}/{}", self.position + 1, self.count)
        } else {
            String::new()
        }
    }
}

/// Per-window cycle cursors.
#[derive(Debug, Default)]
pub struct CycleState {
    cursors: HashMap<u64, (WindowAction, usize)>,
}

impl CycleState {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn can_cycle(action: WindowAction) -> bool {
        chain_for(action).is_some()
    }

    pub fn select(
        &mut self,
        window: u64,
        requested: WindowAction,
        enabled: bool,
    ) -> CycleSelection {
        let Some(chain) = chain_for(requested) else {
            self.cursors.remove(&window);
            return CycleSelection {
                requested,
                effective: requested,
                position: 0,
                count: 0,
                is_cycle: false,
            };
        };
        if !enabled {
            self.cursors.remove(&window);
            return CycleSelection {
                requested,
                effective: requested,
                position: 0,
                count: 0,
                is_cycle: false,
            };
        }
        let requested_pos = chain.iter().position(|a| *a == requested).unwrap_or(0);
        let mut position = requested_pos;
        if let Some((prev_requested, prev_pos)) = self.cursors.get(&window) {
            if *prev_requested == requested {
                position = (prev_pos + 1) % chain.len();
            }
        }
        CycleSelection {
            requested,
            effective: chain[position],
            position,
            count: chain.len(),
            is_cycle: true,
        }
    }

    pub fn commit(&mut self, window: u64, selection: CycleSelection) {
        if selection.is_cycle {
            self.cursors
                .insert(window, (selection.requested, selection.position));
        }
    }

    pub fn forget(&mut self, window: u64) {
        self.cursors.remove(&window);
    }
}

fn chain_for(action: WindowAction) -> Option<&'static [WindowAction; 3]> {
    use WindowAction::*;
    match action {
        LeftHalf | LeftThird | LeftTwoThirds => Some(&[LeftHalf, LeftThird, LeftTwoThirds]),
        RightHalf | RightThird | RightTwoThirds => Some(&[RightHalf, RightThird, RightTwoThirds]),
        TopHalf | TopThird | TopTwoThirds => Some(&[TopHalf, TopThird, TopTwoThirds]),
        BottomHalf | BottomThird | BottomTwoThirds => {
            Some(&[BottomHalf, BottomThird, BottomTwoThirds])
        }
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn non_cyclable_actions_pass_through() {
        let mut state = CycleState::new();
        let sel = state.select(1, WindowAction::Center, true);
        assert!(!sel.is_cycle);
        assert_eq!(sel.effective, WindowAction::Center);
    }

    #[test]
    fn cardinal_actions_cycle_three_steps() {
        let mut state = CycleState::new();
        let first = state.select(7, WindowAction::RightHalf, true);
        assert_eq!(first.effective, WindowAction::RightHalf);
        state.commit(7, first);
        let second = state.select(7, WindowAction::RightHalf, true);
        assert_eq!(second.effective, WindowAction::RightThird);
        state.commit(7, second);
        let third = state.select(7, WindowAction::RightHalf, true);
        assert_eq!(third.effective, WindowAction::RightTwoThirds);
    }

    #[test]
    fn disabled_cycle_clears_cursor() {
        let mut state = CycleState::new();
        let first = state.select(9, WindowAction::LeftHalf, true);
        state.commit(9, first);
        let sel = state.select(9, WindowAction::LeftHalf, false);
        assert!(!sel.is_cycle);
    }
}
